Add GafferScene::MeshSplit

danieldresser-ie opened the pull request:
This feels like a fairly straightforward presentation of IECoreScene::MeshSplitter as a BranchCreator.

Now that 1.2 is released, should I rebase this onto 1.2?

johnhaddon commented on python/GafferSceneTest/MeshSplitTest.py:
```diff
@@ -0,0 +1,293 @@
+		with six.assertRaisesRegex( self, RuntimeError, "Cannot use nameFromSegment with primitive variable uniformIndividualFaces because it is not a string." ) :
+			self.sceneSummary( MeshSplit["out"] )
+			
+
```
Would be good to sprinkle `assertSceneValid( meshSplit["out"] )` through this function to get some coverage for the set and bound computations for free.

johnhaddon reviewed (changes requested):
Thanks Daniel! Comments inline as usual...

johnhaddon commented:
> Now that 1.2 is released, should I rebase this onto 1.2?

Yep, same rule as always - bug fixes onto the oldest maintenance branch, back-compatible features onto the newest, and breaking changes onto `main`.

danieldresser-ie commented on src/GafferScene/MeshSplit.cpp:
```diff
@@ -0,0 +1,470 @@
+		MeshSplitterData( ConstMeshPrimitivePtr mesh, const PrimitiveVariable &primitiveVariable, const IECore::Canceller *canceller, bool useNamesTag ) :
+			m_splitter( mesh, primitiveVariable, m_nameMap, canceller )
+		{
+			InternedStringVectorDataPtr namesData = new InternedStringVectorData;
```
Having m_names be ConstInternedStringVectorDataPtr seems to help communicate that it isn't modified after construction?

johnhaddon replied:
Yep, good point.

danieldresser-ie commented on src/GafferScene/MeshSplit.cpp:
```diff
@@ -0,0 +1,470 @@
+{
+	if( branchPath.size() == 0 )
+	{
+		inPlug()->objectPlug()->hash( h );
```
In addition to moving it to the top, I'm assuming I should also replace hashBranchTransform with hashObjectTransform?  ( I think I must have copied this from a class with different requirements, sorry ).

danieldresser-ie commented on src/GafferScene/MeshSplit.cpp:
```diff
@@ -0,0 +1,470 @@
+	ConstMeshSplitterDataPtr meshSplitter;
+	{
+		ScenePlug::PathScope s( context, &sourcePath );
+		meshSplitter = static_pointer_cast<const MeshSplitterData>( meshSplitterPlug()->getValue() );
```
Hmm, I think there is an argument that if the branchPath size is non-zero, that means that branchChildNames at the parent location must have returned something, which is also driven by the MeshSplitter, so the MeshSplitter must be non-null?  Maybe some more error handling wouldn't hurt though, would you like an assert or an exception?

danieldresser-ie commented on python/GafferSceneUI/MeshSplitUI.py:
```diff
@@ -0,0 +1,78 @@
+	"description",
+	"""
+	Split a mesh into separate meshes for each unique value of a chosen
+	per face primitive variable.  The meshes will be created as children
```
Code uses a monospaced font, so I would actually argue that typesetting principles from a typewriter do in fact apply, at least far more than typesetting principles based on proportional fonts.  But I don't really feel too strongly, it's just a habit, I can try to stop doing it if it bugs you.

danieldresser-ie replied:
I did double check that whatever sort of markdown we're using is ignoring excess spaces - pretty sure it would bug me if there were double spaces showing up in proportional font displays, but I don't think that's happening. Not sure about widely accepted when it specifically comes to monospace typesetting, but I can try to change.  Humourously, this is the one place I agree with PEP8 - it requires double spaces after periods in comments :P

danieldresser-ie commented:
Bit of a last minute upload since I did get back to this til late after leaving work with it not quite done, but I think I have now either commented on or addressed all feedback on both this and the Cortex PR, including the new "preciseBounds" flag with accompanying API ( suggested by Murray to default off ).

johnhaddon reviewed (changes requested):
Thanks Daniel! Could you take care of the remaining small comments at the same time as squashing down please? Oh, and update Changes.md to include this in the Features section.

Spent a bit more time playing about with this today and it has all been working as expected. I did notice one thing that was a bit counterintuitive though. I was generating segment ids based on proximity to a coordinate system, splitting a mesh into concentric rings. So I knew my IDs were all consecutive and increasing from 0 at the centre to N at the edge. But obviously that didn't translate into the child names, because they come from the `0-N` assigned by the MeshSplitter. And that order even jumps around depending on which order the segments are encountered when iterating the faces. And that makes it hard to treat the resulting pieces consistently - my setup started as a mock up of something a lighter asked for - splitting a mesh based on proximity to a coordinate system, and treating the closest mesh differently.

This does all make sense when you think about the general case, but was still a surprise to me at first. Since one of our primary use cases is with MeshSegments, where we again have nicely behaved indices, I wouldn't be surprised if this surprises other folks too. Perhaps we can make `nameFromSegment` work for integer types too?

danieldresser-ie commented:
Updated to work with the new API for IECoreScene::MeshSplitter, and support nameFromSegments for all types ( even silly ones ).

It took me way longer than it should have to figure out which of the ways of invoking fmtlib is the right one here, but the result is simple and works great - this knocked another 20% off the runtime for my test compared to my hacky snprintf, for a total of about 40% reduction compared to stringstream, which feels reasonably worthwhile.

There is an additional commit that addresses the last few bits of feedback.  I've also snuck in a commit that modifies config/ie/options to let me use fmtlib without needing to hack things at IE - haven't really confirmed that with folks yet, but it makes sense, and currently 1.2 just fails to build at IE, so it probably doesn't hurt to commit that.

johnhaddon commented on python/GafferSceneTest/MeshSplitTest.py:
```diff
@@ -321,5 +321,82 @@ def testWithSegment( self ) :
+			IECore.InternedStringVectorData( [ ", ".join( [ "0.3333333333333333" ] * 16 ) ] )
+		)
+
+	@unittest.skipIf( True or GafferTest.inCI(), "Performance not relevant on CI platform" )
```
Remove `True`?

johnhaddon commented on src/GafferScene/MeshSplit.cpp:
```diff
@@ -54,33 +58,192 @@ using namespace IECoreScene;
+	{
+		throw IECore::Exception( "Internal error in formatAsInternedString - buffer not large enough" );
+	}
+	current = fmt::format_to_n( current, end - current, FMT_COMPILE("{}"), a).out;
```
Can we return `result.out` here, instead of calling `format_to_n()` again?

danieldresser-ie replied:
Oh, that's an embarrassing mistake.  Yeah, it was absolutely my intention to use result.out, this was just a last minute fix to add the error handling, and made a copy-paste mistake.  It's interesting that this isn't even that obvious in the perf test - about a 3% slowdown, as opposed to 20% for snprintf or 40% for stringstream - does kinda validate that fmtlib is a good approach.

johnhaddon commented on src/GafferScene/MeshSplit.cpp:
```diff
@@ -54,33 +58,192 @@ using namespace IECoreScene;
+			{
+				for( int i = 0; i < m_splitter.numMeshes(); i++ )
+				{
+					InternedString name( i );
```
Probably worth a canceller check in here?

johnhaddon commented on src/GafferScene/MeshSplit.cpp:
```diff
@@ -54,33 +58,192 @@ using namespace IECoreScene;
+							std::string buffer;
+							for( int i = 0; i < m_splitter.numMeshes(); i++ )
+							{
+								const ElementType& val = m_splitter.value< ElementType >( i );
```
Check canceller?

johnhaddon reviewed (changes requested):
LGTM, with just a couple of small comments it'd be good to address as you squash this down for merging.

danieldresser-ie commented:
Addressed the last few comments, squashed, and added Changes entry.

johnhaddon reviewed (approved):
LGTM. Just needs a Cortex release so we can merge (I've pinged Ivan and Daniel about that, since there are some questions about how we handle the dependencies for 1.2).

danieldresser-ie commented:
Rebased and replaced include guard with pragma once.

johnhaddon commented:
Pushed a rebase to fix merge conflicts and pick up the new dependencies packages with the required `MeshAlgo` features.
